compiler/rustc_lint/src/default_field_always_invalid.rs

```diff
@@ -0,0 +1,91 @@
+use rustc_hir as hir;
+use rustc_middle::lint::LintLevelSource;
+use rustc_middle::mir::interpret::ErrorHandled;
+use rustc_session::lint::Level;
+use rustc_session::{declare_lint, declare_lint_pass};
+
+use crate::lints::DefaultFieldAlwaysInvalidConst;
+use crate::{LateContext, LateLintPass};
+
+declare_lint! {
+    /// The `default_field_always_invalid_const` lint checks for structs with
+    /// default fields const values that will *always* fail to be created.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![feature(default_field_values)]
+    /// #[deny(default_field_always_invalid_const)]
+    /// struct Foo {
+    ///     bar: u8 = 130 + 130, // `260` doesn't fit in `u8`
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Without this lint, the error would only happen only during construction
+    /// of the affected type. For example, given the type above, `Foo { .. }`
+    /// would always fail to build, but `Foo { bar: 0 }` would be accepted. This
+    /// lint will catch accidental cases of const values that would fail to
+    /// compile, but won't detect cases that are only partially evaluated.
+    pub DEFAULT_FIELD_ALWAYS_INVALID_CONST,
+    Deny,
+    "using this default field will always fail to compile"
+}
+
+declare_lint_pass!(DefaultFieldAlwaysInvalid => [DEFAULT_FIELD_ALWAYS_INVALID_CONST]);
+
+impl<'tcx> LateLintPass<'tcx> for DefaultFieldAlwaysInvalid {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx hir::Item<'tcx>) {
+        let data = match item.kind {
+            hir::ItemKind::Struct(data, _generics) => data,
+            _ => return,
+        };
+        let hir::VariantData::Struct { fields, recovered: _ } = data else {
+            return;
+        };
+
+        let (level, source) =
+            cx.tcx.lint_level_at_node(DEFAULT_FIELD_ALWAYS_INVALID_CONST, item.hir_id());
+        match level {
+            Level::Deny | Level::Forbid => {}
+            Level::Warn | Level::ForceWarn(_) | Level::Expect(_) => {
+                // We *can't* turn the const eval error into a warning, so we make it a
+                // warning to not use `#[warn(default_field_always_invalid_const)]`.
```

Comment:
This lint only really has to modes: enabled and disabled. If enabled and the const has an error, we *will* emit an unconditional error, even if set to warn. This is because we do `const_eval_poly` further down, and if there's, lets say, a `panic!()` in the `const` expression that will be an unconditional error. It would be preferable to have a mode where we downgrade those to warnings, or agree that "on or off" lints are ok.

I'm not entirely happy with that but I believe it is a reasonable stop-gap to be addressed ahead of stabilization instead of now.

Reply:
Changed to be an unconditional error now.